schemaregistry/schemaregistry.md

```diff
@@ -0,0 +1,257 @@
+# CNCF Schema Registry API Version 0.1-rc01s
+
+## Abstract
+
+This specification defines a simple API for storing, organizing, and accessing
+schema documents for use with serialization and data validation frameworks such
+as Apache Avro and JSON Schema.
+
+## Status of this document
+
+This document is a working draft.
+
+## 1. Introduction
+
+Schema-dependent serialization formats and related frameworks such as Apache
+Avro are popular in eventing and messaging scenarios. With most of the metadata
+information such as names and types of fields being held externally in schema
+documents, these frameworks are able to produce very compact on-wire
+representations of structured data that is carries as the payload of events and
+messages. While the wire-footprint savings are often very significant, the added
+cost is that of schema handling. All communicating parties who need to
+deserialize the event or message payload need access to the schema used to
+publish it.
+
+For formats that do not depend on external schemas for serialization, schemas
+might still be required in some scenarios to allow for consumers or inspecting
+intermediaries to validate the data structure's compliance with a set of rules.
+
+This specification defines a very simple API focused on storing, organizing, and
+accessing such documents. While there are very sophisticated metadata management
+solutions available as open source und under neutral governance already, the focus
+of this specification is not great sophistication but straightforward simplicity.
+
+This schema registry deals with only three top-level elements:
+
+- Schema Group: A named collection of schemas. Each group holds a logically
+  related set of schemas, typically belonging to a particular application or
+  subject to shared access control rules.
+- Schema: A document describing the structure, names, and types of some
+  structured data payload.
+- Version: A specific version of a schema document. Even though not prescribed
+  in this specification, an implementation might choose to impose compatibility
+  constraints on versions following the initial version of a schema.
+
+> NOTE: There are two fundamentally different options for how we can organize
+> multiple schema formats and this specification draft reflects both: 
+>
+> 1) In an "ideal" REST service, any data entity can be represented ("R" in
```

Comment:
Does a registry have to be REST-ful to begin with? Seems worth spending some cycles on challenging this, e.g. it could be interesting to have a registry which is events-based itself, i.e. it'd publish streams of schema versions. That'd put more emphasis on enabling interested consumers to keep copies of schemas they're interested themselves in a push-based manner, instead of relying on a pull-based model.

Reply:
They could optionally support the Discovery and Subscription API specs!  :-)

Reply:
@gunnarmorling You ultimately need a place where you can just go and look things up. A serialization framework that turns an object graph into a serialized Avro container needs to be able to fetch the correct schema in the current state of the art. This interface is literally for interoperable access to the place where the copies of the schemas are kept such that it can be accessed from multiple language runtimes and across the network. Schema distribution is a different concern, IMO, and because of the significant size of many schemas, that interaction is best done using a claim-check pattern where a schema version update gets announced and then the schema is fetched by interested parties. I do like the idea of standardizing those notifications, though; I will add change events that will let you do what you are looking for.